File: Algorithm/programmers/pg_호텔 방 배정.py
# union-find | programmers 호텔 방 배정
# github.com/chj3748
import sys


def input():
    return sys.stdin.readline().rstrip()


import sys

sys.setrecursionlimit(200010)


def solution(k, room_number):
    from collections import defaultdict
    rooms = defaultdict(int)
    answer = []

    def find_r(x):
        if not rooms[x]:
            rooms[x] = x + 1
            return x
        temp = find_r(rooms[x])
        rooms[x] = temp + 1
        return temp

    for r in room_number:
        answer.append(find_r(r))
    return answer

File: Algorithm/programmers/pg_호텔 방 배정.cpp
// union-find | programmers 호텔 방 배정
// github.com/chj3748

#include <string>
#include <vector>
#include <map>
#define ll long long
using namespace std;
map<ll, ll> m;
ll findset(ll x)
{
	if (m[x] == 0) return x;
	return m[x] = findset(m[x]);
}

vector<long long> solution(long long k, vector<long long> room_number) {
	vector<long long> answer;
	for (int i = 0; i < room_number.size(); i++)
	{
		ll now = room_number[i];
        ll nowEnd = findset(now);
        m[nowEnd] = nowEnd + 1;
        answer.push_back(nowEnd);

	}
	return answer;
}